### PR TITLE
ControlledQubitUnitary inherits from Controlled

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
ControlledQubitUnitary was defined as a subclass of `QubitUnitary`, and implemented a variety of methods that are duplicated elsewhere.

**Description of the Change:**
- Reorganize so that `ControlledQubitUnitary` inherits from the Controlled class.
- Add broadcasting to the `Controlled` class. 
- Move `ControlledQubitUnitary` from `matrix_ops.py` to `controlled_ops.py` (avoids circular import). 

**Benefits:**
Less duplicate code. Inherits functionality it did not previously have from `Controlled` (methods `decomposition`, `expand` and `sparse_matrix` now return a result instead of raising an error). Future operators inheriting from `Controlled` will support parameter broadcasting as long as their base does.

**Possible Drawbacks:**
The hyperparameters dictionary now has the same format as other Controlled operators, so `op.hyperparameters['u_wires']` is no longer an option to access the wires the unitary matrix operates on. This information can be accessed via `op.base.wires` or `op.target_wires` instead. Otherwise operator behaviour should be the same as before reorganizing, so there shouldn't be any other issues.

### CODE REVIEW

File: pennylane/ops/op_math/controlled_class.py [hunk 1]
```diff
@@ -414,33 +418,118 @@ def has_adjoint(self):
...
-        return Controlled(
+        return cls(
             base=self.base.simplify(),
             control_wires=self.control_wires,
             control_values=self.control_values,
             work_wires=self.work_wires,
         )
 
+    @staticmethod
+    def compute_matrix(
```
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_class.py:460]: In many situations, `compute_matrix` is easier to define. It also allows us to add `@lru_cache` in non-parametric cases.

But we don't really *need* `compute_matrix` to be defined. The rest of PennyLane just depends on the `matrix` method.  So I don't think we particularly still need this static method.

Comment (lillian542) [reply pennylane/ops/op_math/controlled_class.py]: Removed

**Comment by lillian542** [comment]
Note: all the tests removed from `tests/ops/qubit/test_matrix_ops.py` were copy-pasted into `tests/ops/op_math/test_controlled_ops.py`. The tests were moved to reflect the new location of the operator they test (which had to be moved from the `qubit` folder into a file in the `math_op` folder because importing `Controlled` in the `qubit` folder created circular imports.

File: pennylane/__init__.py [hunk 2]
```diff
@@ -65,7 +65,7 @@
     shadow_expval,
 )
 from pennylane.ops import *
-from pennylane.ops import adjoint, ctrl, exp, op_sum, pow, prod, s_prod
+from pennylane.ops import adjoint, ctrl, exp, op_sum, pow, prod, s_prod, ControlledQubitUnitary
```
Comment (AlbertMitjans) [inline pennylane/__init__.py:68]: Do we really want to import this top-level?

File: pennylane/devices/tests/test_gates.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 # pylint: disable=no-self-use
 # pylint: disable=too-many-arguments
 # pylint: disable=pointless-statement
+# pylint: disable=unnecessary-lambda-assignment
```
Comment (AlbertMitjans) [inline pennylane/devices/tests/test_gates.py:21]: Why not solving the issue instead? You just need to change:
```python
my_func = lambda x: x
```
by
```
def my_func(x):
  return x
```
wherever needed.

Comment (AlbertMitjans) [reply pennylane/devices/tests/test_gates.py:21]: Just realised this is a test file, so it's not that important. But we should push towards using less `# pylint: disable` statements.

File: pennylane/ops/op_math/controlled_class.py [hunk 4]
```diff
@@ -166,6 +166,9 @@ def __new__(cls, base, *_, **__):
     def __init__(
         self, base, control_wires, control_values=None, work_wires=None, do_queue=True, id=None
     ):
+        if control_wires is None:
+            raise ValueError("Must specify control wires")
+
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:171]: Why this check? Did you find a corner case where `control_wires=None`?

File: pennylane/ops/op_math/controlled_class.py [hunk 5]
```diff
@@ -177,6 +180,10 @@ def __init__(
                     "Specifying control values as a string is deprecated. Please use Sequence[Bool]",
                     UserWarning,
                 )
+
+                if not set(control_values).issubset({"0", "1"}):
+                    raise ValueError("String of control values can contain only '0' or '1'.")
+
                 control_values = [(x == "1") for x in control_values]
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:187]: Again the eternal discussion of whether to have a lot of input checks and less performance, or give more freedom to the user to improve performance 😆 

In this case I would not add this check, given that in python casting to bool any number different than 0 returns `True`:
```pycon
>>> bool(2)
True
```

Thus we can do the following to keep with python standards:
```suggestion
                control_values = [(x != "0") for x in control_values]
```

File: pennylane/ops/op_math/controlled_class.py [hunk 6]
```diff
@@ -277,8 +296,9 @@ def map_wires(self, wire_map: dict):
         new_base = self.base.map_wires(wire_map=wire_map)
         new_control_wires = Wires([wire_map.get(wire, wire) for wire in self.control_wires])
         new_work_wires = Wires([wire_map.get(wire, wire) for wire in self.work_wires])
+        cls = self.__class__
 
-        return Controlled(
+        return cls(
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:301]: Cosmetic comment: you can call `self.__class__` directly.

File: pennylane/ops/op_math/controlled_class.py [hunk 7]
```diff
@@ -414,27 +437,38 @@ def has_adjoint(self):
         return self.base.has_adjoint
 
     def adjoint(self):
-        return Controlled(
-            self.base.adjoint(), self.control_wires, self.control_values, self.work_wires
+        cls = self.__class__
+        return cls(
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:441]: Same here

File: pennylane/ops/op_math/controlled_class.py [hunk 8]
```diff
@@ -414,27 +437,38 @@ def has_adjoint(self):
...
+            control_values=self.control_values,
+            work_wires=self.work_wires,
         )
 
     def pow(self, z):
         base_pow = self.base.pow(z)
+        cls = self.__class__
         return [
-            Controlled(op, self.control_wires, self.control_values, self.work_wires)
+            cls(
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:452]: Same here

File: pennylane/ops/op_math/controlled_class.py [hunk 9]
```diff
@@ -414,27 +437,38 @@ def has_adjoint(self):
...
+                op,
+                self.control_wires,
+                control_values=self.control_values,
+                work_wires=self.work_wires,
+            )
             for op in base_pow
         ]
 
     def simplify(self) -> "Controlled":
+        cls = self.__class__
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_class.py:462]: Same here

File: tests/ops/op_math/test_controlled_op.py [hunk 10]
```diff
@@ -649,14 +650,16 @@ def test_do_queue_false(self):
...
         x = np.array([1.0, 2.0, 3.0])
         base = qml.RX(x, 0)
         op = Controlled(base, 1)
 
-        with pytest.raises(qml.operation.MatrixUndefinedError):
-            op.matrix()
+        matrix = op.matrix()
+
+        assert matrix.shape == (3, 4, 4)
+        assert matrix.shape[-2:] == (4, 4)
```
Comment (AlbertMitjans) [inline tests/ops/op_math/test_controlled_op.py:662]: Why adding this duplicated test?

File: tests/ops/qubit/test_attributes.py [hunk 11]
```diff
@@ -436,3 +435,149 @@ def test_qaoa_embedding(self, features, weights, num_wires, batch_size):
             features, weights, wires=list(range(num_wires)), local_field=qml.RY
         )
         op.decomposition()
+
+
+class TestControlledSupportsBroadcasting:
```
Comment (AlbertMitjans) [inline tests/ops/qubit/test_attributes.py:440]: Nice!! 🚀 

I realised there are some tests from `TestSupportsBroadcasting` that are not being tested here. Is it because their controlled version is not supported?

Comment (AlbertMitjans) [reply tests/ops/qubit/test_attributes.py:440]: Also, should we move this to `test_controlled_op.py`?

**Review by AlbertMitjans — COMMENTED** [review]
Good job!! 💯 🚀  Some minor comments:

1. Please try to separate "different" tasks in different PRs. Especially when moving code around: if you create a PR where you ONLY move the code you will save the reviewer a lot of time.
2. I assumed that the `ControlledQubitUnitary` class and tests were ONLY moved, thus I didn't check its code.

File: pennylane/__init__.py (hunk 2 above)
Comment (lillian542) [reply pennylane/__init__.py:68]: Yes, but maybe not right here. It is available as `qml.ControlledQubitUnitary` on master and should remain so following these changes. Moving it from `qubit.matrix_ops` broke using `qml.ControlledQubitUnitary`, and putting the import here seemed like the easiest way to make sure it was available. But maybe it would be better to put it elsewhere?

File: pennylane/devices/tests/test_gates.py (hunk 3 above)
Comment (lillian542) [reply pennylane/devices/tests/test_gates.py:21]: The pylint `disable` statement is not related to anything I changed, I only updated the line `all_available_ops.add("ControlledQubitUnitary")`. It's related to a section further down that was added over a year ago; I have `pre-commit` installed and pylint insisted I add this statement before committing my change, even though it relates to an unmodified section of the code. 

Since it's a test file and not related to any changes in this PR, I'm going to leave as is.

File: pennylane/__init__.py (hunk 2 above)
Comment (AlbertMitjans) [reply pennylane/__init__.py:68]: I didn't know it was available as `qml.ControlledQubitUnitary`! Then we can leave it as it is.

File: pennylane/ops/op_math/controlled_class.py (hunk 4 above)
Comment (lillian542) [reply pennylane/ops/op_math/controlled_class.py:171]: > 1. Please try to separate "different" tasks in different PRs. Especially when moving code around: if you create a PR where you ONLY move the code you will save the reviewer a lot of time.
> 2. I assumed that the `ControlledQubitUnitary` class and tests were ONLY moved, thus I didn't check its code.

Sorry, this did get a bit messy. I guess I should clarify that the primary purpose of this PR was for ControlledQubitUnitary to inherit as much behaviour as possible from Controlled (instead of having it defined in ControlledQubitUnitary), so the primary changes being made here are changing Controlled and ControlledQubitUnitary. So its definitely **not** the case that the ControlledQubitUnitary class was only moved. 

Moving the code between files was an unfortunate necessity to avoid circular imports between `math_ops` and `qubit folders`. I'll aim to make things like that a separate initial PR in the future.

Adding parameter broadcasting is also a subtask that was needed to make sure that ControlledQubitUnitary didn't lose any features when being modified to inherit its behaviour from Controlled. Here are the changes specific to adding broadcasting: [#3577](https://github.com/PennyLaneAI/pennylane/pull/3577). This probably also should have been a separate PR if possible.

File: tests/ops/op_math/test_controlled_op.py (hunk 10 above)
Comment (lillian542) [reply tests/ops/op_math/test_controlled_op.py:662]: Hahaha... I have no idea why I did that. Will remove.

File: pennylane/ops/op_math/controlled_class.py (hunk 6 above)
Comment (lillian542) [reply pennylane/ops/op_math/controlled_class.py:297]: Updated these

File: tests/ops/qubit/test_attributes.py (hunk 11 above)
Comment (lillian542) [reply tests/ops/qubit/test_attributes.py:440]: I didn't think we would want to do a `Controlled` version of embeddings or state vectors, so I didn't test them - but that might be wrong. I just tried them and they work. Should they be included in the tests?

I think moving them would make sense.

File: pennylane/ops/op_math/controlled_class.py (hunk 4 above)
Comment (albi3ro) [reply pennylane/ops/op_math/controlled_class.py:171]:
```pycon
>>> qml.ops.Controlled(qml.PauliX(0))
TypeError: ControlledOp.__init__() missing 1 required positional argument: 'control_wires'
```
`ControlledQubitUnitary` had that check because the call signature had a default value for `control_wires`. We no longer have a default value, because it is an argument instead of a keyword argument.

I'm Ok with just turning a `ValueError` into a `TypeError` and getting rid of this logic.

File: pennylane/ops/op_math/controlled_class.py [hunk 12]
```diff
@@ -177,7 +180,7 @@ def __init__(
                     "Specifying control values as a string is deprecated. Please use Sequence[Bool]",
                     UserWarning,
                 )
-                control_values = [(x == "1") for x in control_values]
+                control_values = [(x != "0") for x in control_values]
```
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_class.py:183]: Why this change?

File: pennylane/ops/qubit/matrix_ops.py [hunk 13]
```diff
@@ -74,11 +74,12 @@ class QubitUnitary(Operation):
     def __init__(
         self, U, wires, do_queue=True, id=None, unitary_check=False
     ):  # pylint: disable=too-many-arguments
-        wires = Wires(wires)
-
         # For pure QubitUnitary operations (not controlled), check that the number
         # of wires fits the dimensions of the matrix
-        if not isinstance(self, ControlledQubitUnitary):
+
+        if not isinstance(self, qml.ControlledQubitUnitary):
```
Comment (albi3ro) [inline pennylane/ops/qubit/matrix_ops.py:80]: Since we've gotten rid of the inheritance between `QubitUnitary` and `ControlledQubitUnitary`, we can get rid of this check and un-indent all the following code.

File: pennylane/ops/op_math/controlled_class.py (hunk 12 above)
Comment (lillian542) [reply pennylane/ops/op_math/controlled_class.py:183]: See Albert's comment above (just marked as unresolved, so should be visible now)

File: tests/ops/qubit/test_attributes.py (hunk 11 above)
Comment (AlbertMitjans) [reply tests/ops/qubit/test_attributes.py:440]: Adding a test doesn't hurt, thus I would include them.

Let's also move the tests then!

File: pennylane/ops/op_math/controlled_ops.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+    # pylint: disable= too-many-arguments
+    def __init__(
+        self,
+        base,
+        control_wires=None,
+        wires=None,
+        control_values=None,
+        do_queue=True,
+        unitary_check=False,
+        work_wires=None,
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_ops.py:102]: Not used

File: pennylane/ops/op_math/controlled_ops.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+    * Number of parameters: 1
+    * Number of dimensions per parameter: (2,)
+    * Gradient recipe: None
+
+    Args:
+        U (Union[array[complex], QubitUnitary]): square unitary matrix or a QubitUnitary operation
+        control_wires (Union[Wires, Sequence[int], or int]): the control wire(s)
+        wires (Union[Wires, Sequence[int], or int]): the wire(s) the unitary acts on (optional if U is provided as a QubitUnitary)
+        control_values (List[int, bool]): a list providing the state of the control qubits to control on (default is the all 1s state)
+        unitary_check (bool): whether to check whether an array U is unitary when creating the operator (default False)
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_ops.py:50]: `do_queue` missing

File: pennylane/ops/op_math/controlled_ops.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+    **Example**
+
+    The following shows how a single-qubit unitary can be applied to wire ``2`` with control on
+    both wires ``0`` and ``1``:
+
+    >>> U = np.array([[ 0.94877869,  0.31594146], [-0.31594146,  0.94877869]])
+    >>> qml.ControlledQubitUnitary(U, control_wires=[0, 1], wires=2)
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_ops.py:58]: Add output

File: pennylane/ops/op_math/controlled_ops.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+    Alternatively, the same operator can be constructed with a QubitUnitary:
+
+    >>> base = qml.QubitUnitary(U, wires=2)
+    >>> qml.ControlledQubitUnitary(base, control_wires=[0, 1])
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_ops.py:63]: Add output

File: pennylane/ops/op_math/controlled_ops.py [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+        self._name = "ControlledQubitUnitary"
+
+    def _controlled(self, wire):
+        ctrl_wires = self.control_wires + wire
+        values = None if self.control_values is None else self.control_values + [True]
+        new_op = ControlledQubitUnitary(
+            self.base,
+            control_wires=ctrl_wires,
+            control_values=values,
+            work_wires=self.work_wires,
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_ops.py:122]: `self.work_wires` is not defined

**Review by AlbertMitjans — COMMENTED** [review]
Some comments on the `ControlledQubitUnitary` class

**Comment by AlbertMitjans** [comment]
[sc-31243]

File: pennylane/ops/op_math/controlled_class.py (hunk 12 above)
Comment (albi3ro) [reply pennylane/ops/op_math/controlled_class.py:183]: Ah, so everything that's not `"0"` gets cast to `True`. Might be good to add a comment as to that behaviour, as this could be a potential source of confusion.

File: pennylane/ops/op_math/controlled_ops.py (hunk 14 above)
Comment (albi3ro) [reply pennylane/ops/op_math/controlled_ops.py:102]: Could just make sure it gets passed to `super().__init__`.  While we don't use work wires much right now outside of `MultiControlledX`, decomposition algorithms will need it when we start having them.

File: pennylane/ops/op_math/controlled_ops.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+        )
+        return new_op.inv() if self.inverse else new_op
```
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_ops.py:124]: Given I'm trying to get rid of in-place inversion, we should just go ahead and get rid of it here and now.

```suggestion
        return new_op
```

File: pennylane/ops/op_math/controlled_ops.py [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+    **Details:**
+
+    * Number of wires: Any (the operation can act on any number of wires)
+    * Number of parameters: 1
+    * Number of dimensions per parameter: (2,)
+    * Gradient recipe: None
+
+    Args:
+        U (Union[array[complex], QubitUnitary]): square unitary matrix or a QubitUnitary operation
```
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/ops/op_math/controlled_ops.py:46]: It does if initialized using a `qml.QubitUnitary`. You can also pass it a matrix, which it will then use to create a `QubitUnitary` base. 

Passing a matrix `U` is the original behaviour, the option to directly pass it a base operator has been added in this PR to mimic the behaviour of SymbolicOp and CompositeOp, since I imagine this is what people would expect if they are used to those operators. But if people have been using it in the past, they would expect to pass just a matrix.

I'll expand a bit in the docstring to make this clearer.

File: pennylane/ops/op_math/controlled_ops.py (hunk 15 above)
Comment (lillian542) [reply pennylane/ops/op_math/controlled_ops.py:50]: Thanks - also in the Controlled docstring :) Updated now!

File: pennylane/ops/op_math/controlled_ops.py (hunk 14 above)
Comment (lillian542) [reply pennylane/ops/op_math/controlled_ops.py:102]: Thanks, yes, it's meant to be passed to the `super().__init__` function. Updated now.

File: pennylane/ops/op_math/controlled_ops.py (hunk 18 above)
Comment (lillian542) [reply pennylane/ops/op_math/controlled_ops.py:122]: It was defaulting to `None` in all instances, since it wasn't passed to the `super.__init__` (just fixed). But `self.work_wires` is defined for all operators that inherit from Controlled.

File: tests/ops/qubit/test_attributes.py (hunk 11 above)
Comment (lillian542) [reply tests/ops/qubit/test_attributes.py:440]: I'm a little confused about why the `Embedding` and `StateVector` tests all include something along the lines of 

```
qml.QubitStateVector.compute_decomposition(state, list(range(num_wires)))
op.decomposition()
```

I'm not sure what that is testing, or whether I should include something similar in the broadcasting tests for the controlled version. Thoughts?

File: pennylane/ops/op_math/__init__.py [hunk 21]
```diff
@@ -43,6 +43,7 @@
     ~CompositeOp
     ~Controlled
     ~ControlledOp
+    ~ControlledQubitUnitary
```
Comment (albi3ro) [inline pennylane/ops/op_math/__init__.py:46]: I don't necessarily think we need to include it here, but maybe that's my preconceived notions of ControlledQubitUnitary as a "normal" operation. My intuition says to remove it and keep it in the same category as "normal" ops, but I'm fine either way.

Comment (lillian542) [reply pennylane/ops/op_math/__init__.py:46]: So would you be inclined to just not list it here, even though it lives in op_math?

Comment (albi3ro) [reply pennylane/ops/op_math/__init__.py:46]: Maybe we can add a new section for operations like `ControlledQubitUnitary`? Especially since we can repeat this restructuring for more "basic" operations.

**Comment by lillian542** [comment]
> It automatically imports everything in the `__all__` dunder list. I think it would be best to keep "ControlledQubitUnitary" in that list, and potentially even in `_qubit__ops__` as well.

Okay, thanks - I didn't know how that worked! 

I rearranged so there is a set in `ops.op_math.__init__` for this and any future `Controlled` ops, which will all then be included in `_qubit__ops__`,  ` _qubit__all__`, and the `__all__` dunder list for `ops`.

File: pennylane/ops/__init__.py [hunk 22]
```diff
@@ -29,16 +29,22 @@
...
+from .qubit import __all__ as _qubit__all__
+from .qubit import ops as _qubit__ops__
+from .qubit import obs as _qubit__obs__
+
+# we would like these to just live in .qubit, but can't because of circular imports
+from .op_math import controlled_qubit_ops as _controlled_qubit__ops__
+
+_qubit__ops__ = _qubit__ops__ | _controlled_qubit__ops__
+_qubit__all__ = _qubit__all__ + list(_controlled_qubit__ops__)
+
```
Comment (AlbertMitjans) [inline pennylane/ops/__init__.py:47]: Why not just import `ControlledQubitUnitary` in the top `__init__.py` file instead?

File: pennylane/ops/op_math/controlled_ops.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (AlbertMitjans) [reply pennylane/ops/op_math/controlled_ops.py]: Thanks for expanding the docstring. But the name of the `arg` should be the same as the argument in the `__init__.py` method. Thus this should be `base`.

File: tests/ops/qubit/test_attributes.py (hunk 11 above)
Comment (AlbertMitjans) [reply tests/ops/qubit/test_attributes.py]: I believe it is testing that `compute_decomposition` doesn't raise any errors when using batching

Comment (AlbertMitjans) [reply tests/ops/qubit/test_attributes.py]: I would also include it in the broadcasting tests of the controlled version

File: pennylane/ops/op_math/controlled_ops.py [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+        super().__init__(
+            base,
+            control_wires,
+            control_values=control_values,
+            work_wires=work_wires,
+            do_queue=do_queue,
+        )
+        self._name = "ControlledQubitUnitary"
+
+    def _controlled(self, wire):
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_ops.py:129]: What is this for? Where is it used?

File: doc/releases/changelog-dev.md [hunk 24]
```diff
@@ -141,9 +149,14 @@
 
  <h3>Breaking changes</h3>
 
+* The target wires of the unitary for `ControlledQubitUnitary` are no longer available via `op.hyperparameters["u_wires"]`.
```
Comment (AlbertMitjans) [inline doc/releases/changelog-dev.md:152]: Ask @trbromley if we should add a deprecation cycle for this

**Review by AlbertMitjans — COMMENTED** [review]
Should we add tests for the `decomposition`, `expand` and `sparse_matrix` methods of the `ControlledQubitUnitary` class?

File: doc/releases/changelog-dev.md (hunk 24 above)
Comment (albi3ro) [reply doc/releases/changelog-dev.md:155]: Fairly internal edge case in my mind.

File: pennylane/ops/op_math/__init__.py [hunk 25]
```diff
@@ -51,15 +51,25 @@
     ~SProd
     ~SymbolicOp
 
+Controlled Operator Classes
+~~~~~~~~~~~~~~~~
```
Comment (albi3ro) [inline pennylane/ops/op_math/__init__.py:55]:
```suggestion
Controlled Operator Classes
~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

`/github/workspace/pennylane/ops/op_math/__init__.py:docstring of pennylane.ops.op_math:41: WARNING: Title underline too short.`

Sphinx is weird.

File: pennylane/ops/op_math/controlled_ops.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply pennylane/ops/op_math/controlled_ops.py]: Ah, thanks! I hadn't thought about that :)

File: pennylane/ops/op_math/__init__.py [hunk 26]
```diff
@@ -51,15 +51,25 @@
     ~SProd
     ~SymbolicOp
 
+Controlled Operator Classes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    ~ControlledQubitUnitary
+
```
Comment (albi3ro) [inline pennylane/ops/op_math/__init__.py:57]:
```suggestion

.. autosummary::
    :toctree: api

    ~ControlledQubitUnitary

```

**Review by albi3ro — APPROVED** [review]
Approved under the condition that the `test_controlled_ops.py` is linted. 

Great to finally have this simplification happening!
